exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/FragmentsRunner.java

```
@@ -0,0 +1,439 @@
+  private PlanFragment rootPlanFragment;
+  private FragmentRoot rootOperator;
+
+  public FragmentsRunner(WorkerBee bee, UserClientConnection initiatingClient, DrillbitContext drillbitContext, Foreman foreman) {
+    this.bee = bee;
+    this.initiatingClient = initiatingClient;
+    this.drillbitContext = drillbitContext;
+    this.foreman = foreman;
+  }
+
+  public WorkerBee getBee() {
+    return bee;
+  }
+
+  public void setPlanFragments(List<PlanFragment> planFragments) {
+    this.planFragments = planFragments;
+  }
+
+  public void setRootPlanFragment(PlanFragment rootPlanFragment) {
+    this.rootPlanFragment = rootPlanFragment;
+  }
+
+  public void setRootOperator(FragmentRoot rootOperator) {
+    this.rootOperator = rootOperator;
+  }
+
+  /**
+   * Submits root and non-root fragments fragments for running.
+   * In case of success move query to the running state.
+   */
+  public void submit() {
+    try {
+      assert planFragments != null;
+      assert rootPlanFragment != null;
+      assert rootOperator != null;
+
+      QueryId queryId = foreman.getQueryId();
+      assert queryId == rootPlanFragment.getHandle().getQueryId();
+
+      QueryManager queryManager = foreman.getQueryManager();
+
+      try {
+        drillbitContext.getWorkBus().addFragmentStatusListener(queryId, queryManager.getFragmentStatusListener());
+        drillbitContext.getClusterCoordinator().addDrillbitStatusListener(queryManager.getDrillbitStatusListener());
+
+        logger.debug("Submitting fragments to run.");
+        // set up the root fragment first so we'll have incoming buffers available.
+        setupRootFragment(rootPlanFragment, rootOperator);
+        setupNonRootFragments(planFragments);
+
+      } catch (ExecutionSetupException e) {
+        foreman.moveToState(QueryState.FAILED, e);
+      }
+
+      foreman.moveToState(QueryState.RUNNING, null);
+      logger.debug("Fragments running.");
+    } finally {
+      foreman.startProcessingEvents();
```

If we get here, we may have bypassed a state transition if an error occurred. Should the above two nested `try` blocks be collapsed so we are sure to catch all exceptions?